### PR TITLE
Argument Completers for FailoverClusters module

### PR DESCRIPTION
**Pull request by roggenk**
Failover Clustering argument completion for parameters
- Cluster (-Cluster, -Name)
- Cluster Resource (Name, VMId)
- Cluster Resource Type
- Cluster Group (Name, VMId)
- Cluster Node (Name)
- Cluster Shared Volume (Name)
- Cluster Storage
- Cluster Network (Name)
- Cluster Network Interface (Name)
- Cluster Virtual Machine Role (Name, VMId)

### CODE REVIEW

**Comment by lzybkr** [comment]
Thanks for all the contributions.

Can you make sure to do a few things in any more PRs you send:
* Name the file <Module>.ArgumentCompleters.ps1 (this PR looks fine, but you didn't do that in others)
* Add the file that you're adding to Install.ps1
* Let me know if any of the completions are slow e.g. > 250ms - sometimes it's better to return w/ no completions but build a cache the first time so we don't block editing.

**Comment by roggenk** [comment]
Hi Jason,

No problem.

This module indeed has some slow completers:
- All registrations using function FailoverClusters_ClusterResourceNameArgumentCompletion 
Example: Get-ClusterResource -Name
- All registrations using function FailoverClusters_ClusterResourceVMIdArgumentCompletion
Example: Get-ClusterResource -VMId 

Does that helps?
Question: How do you build a cache first?

Kurt

**Comment by roggenk** [comment]
Hi Jason,
Is there soms info you are missing from me?

Thans,
Kurt

File: FailoverClusters.ArgumentCompleters.ps1 [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+{
+    param($commandName, $parameterName, $wordToComplete, $commandAst, $fakeBoundParameter)
+
+    $OptionalCluster = @{}
+    $Cluster = $fakeBoundParameter["Cluster"]
+    if($Cluster)
+    {
+        $OptionalCluster.Cluster = $Cluster
+    }
+    # These completions are slow
```
Comment (roggenk) [inline FailoverClusters.ArgumentCompleters.ps1]: Slow completer.

Comment (roggenk) [inline FailoverClusters.ArgumentCompleters.ps1]: Slow completer.

File: FailoverClusters.ArgumentCompleters.ps1 [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
```
Comment (roggenk) [inline FailoverClusters.ArgumentCompleters.ps1]: Slow completer.

File: FailoverClusters.ArgumentCompleters.ps1 [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
```
Comment (roggenk) [inline FailoverClusters.ArgumentCompleters.ps1]: Slow completer.

File: FailoverClusters.ArgumentCompleters.ps1 [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
```
Comment (roggenk) [inline FailoverClusters.ArgumentCompleters.ps1]: Slow completer.

File: FailoverClusters.ArgumentCompleters.ps1 [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
```
Comment (roggenk) [inline FailoverClusters.ArgumentCompleters.ps1]: Slow completer.

File: FailoverClusters.ArgumentCompleters.ps1 [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
```
Comment (roggenk) [inline FailoverClusters.ArgumentCompleters.ps1]: Slow completer.

File: FailoverClusters.ArgumentCompleters.ps1 [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
```
Comment (roggenk) [inline FailoverClusters.ArgumentCompleters.ps1]: Slow completer.

File: FailoverClusters.ArgumentCompleters.ps1 [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
```
Comment (roggenk) [inline FailoverClusters.ArgumentCompleters.ps1]: Slow completer.

**Comment by roggenk** [comment]
Hi Jason,
I believe all requested information is there.
Could you have a look?

Thanks,
Kurt

**Pull request merged**